dashboard/lib/pd/foorm/foorm_parser.rb

```diff
@@ -0,0 +1,121 @@
+# Parses Foorm forms into a useful format for looking up questions by form name, form version
+# and question name
+# Format is:
+# {
+#   <form-name>.<form-version: {
+#     <question_name>: {
+#       title: "sample title",
+#       type: "text/singleSelect/multiSelect/matrix/scale",
+#       # for singleSelect/multiSelect/scale
+#       choices: {
+#         choice_1_name: "choice 1 value",
+#         ...
+#       },
+#       # if has other choice
+#       other_text: "Other choice text",
+#       # for matrix
+#       rows: {
+#         row_1_name: "row 1 value",
+#         ...
+#       },
+#       columns: {
+##         column_1_name: "column 1 value",
+##         ...
+##       }
+#     }
+#   }
+# }
+module Pd::Foorm
+  class FoormParser
+    include Constants
+    extend Helper
+
+    # parse all forms in given list and return object in format above
+    def self.parse_forms(forms)
+      parsed_forms = {}
+      forms.each do |form|
+        parsed_form = {}
+        form_questions = JSON.parse(form.questions, symbolize_names: true)
+        form_questions[:pages].each do |page|
+          page[:elements].each do |question_data|
+            parsed_form = parsed_form.merge(parse_element(question_data))
+          end
+        end
+        parsed_forms[get_form_key(form.name, form.version)] = parsed_form
+      end
+      parsed_forms
+    end
+
+    # parse a form element and return hash of {question_name->question_data,...}
+    # Form element may be a panel which contains questions, therefore resulting hash
+    # may contain one or more questions
+    def self.parse_element(question_data)
+      parsed_questions = {}
+      if question_data[:type] == 'panel'
+        question_data[:elements].each do |panel_question_data|
+          parsed_questions = parsed_questions.merge(parse_element(panel_question_data))
```

Comment:
Can this be `merge!`?

Reply:
ah yes it can. do we generally prefer using the ! version for things like this?

Reply:
I get the impression that in Ruby, you look for the most compact version of code.  So long as it remains comprehensible, of course!